### PR TITLE
plugin/file: simplify locking

### LINKED ISSUE #2975: plugin/file query deadlock
When we running coredns using file plugin for a while, the coredns did'nt reponse any query served by file plugin,meanwhile, the k8s plugin still working. here are some infomation, I killed the coredns process with ABRT signal and find lots of goroutine stuck at `plugin/file/lookup.go:36`, is `z.reloadMu.RLock()`.

the coredns version we use: 
```
# /usr/bin/coredns -version
CoreDNS-1.5.2
linux/amd64, go1.12.6, d933f63
``` 
the corefile:
```
nvsdbl.example.com. {
    kubernetes {
        endpoint https://nvsdbl.k8.example.com:6443
        tls /etc/coredns/k8s-client.cert /etc/coredns/k8s-client.key /etc/coredns/k8s-ca.cert
        ttl 30
    }
    log
    prometheus 0.0.0.0:9153
}

example.com {
    file /etc/coredns/example.com.signed {
        reload 5s
    }
    log
    prometheus 0.0.0.0:9153
}

```

when we use `queryperf` to benchmark coredns, about 5 or 10 seconds the coredns will get stuck. the command and ouput is:
```
# ./queryperf -q 100 -d x.t -s 10.13.2.35 -l 1000                                                                                                                             

DNS Query Performance Testing Tool
Version: $Id: queryperf.c,v 1.12 2007/09/05 07:36:04 marka Exp $
[Status] Processing input data
[Status] Sending queries (beginning with 10.13.2.35)
[Timeout] Query timed out: msg id 47458
[Timeout] Query timed out: msg id 47550
[Timeout] Query timed out: msg id 47557
[Timeout] Query timed out: msg id 47564
[Timeout] Query timed out: msg id 47625
[Timeout] Query timed out: msg id 47640
[Timeout] Query timed out: msg id 47650
[Timeout] Query timed out: msg id 47651
[Timeout] Query timed out: msg id 47656
[Timeout] Query timed out: msg id 47657
[Timeout] Query timed out: msg id 47706
[Timeout] Query timed out: msg id 47707
[Timeout] Query timed out: msg id 47719
[Timeout] Query timed out: msg id 47723
[Timeout] Query timed out: msg id 47732
[Timeout] Query timed out: msg id 47735
[Timeout] Query timed out: msg id 47739
[Timeout] Query timed out: msg id 47740
```

and the shorten dumped stack info:
```
coredns[23120]: goroutine 175906 [semacquire, 1 minutes]:
coredns[23120]: sync.runtime_SemacquireMutex(0xc00063838c, 0x4bdf00)
coredns[23120]: /tmp/tmp.nhEpihdcHr/go/src/runtime/sema.go:71 +0x3d
coredns[23120]: sync.(*RWMutex).RLock(...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/go/src/sync/rwmutex.go:50
coredns[23120]: github.com/coredns/coredns/plugin/file.(*Zone).Lookup(0xc000638280, 0x1cfd680, 0xc001cd8a50, 0xc002274a20, 0x1d1ce80, 0xc0011b1100, 0x0, 0x0, 0x0, 0xc001ae60dc, ...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/file/lookup.go:36 +0x3202
coredns[23120]: github.com/coredns/coredns/plugin/file.File.ServeDNS(0x0, 0x0, 0xc000696720, 0xc00035b0f0, 0x1, 0x1, 0x1cfd680, 0xc001cd8a50, 0x1d1ce80, 0xc0011b1100, ...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/file/file.go:82 +0x29e
coredns[23120]: github.com/coredns/coredns/plugin.NextOrFailure(0x19c5c23, 0x3, 0x1cd9c40, 0xc000426f60, 0x1cfd680, 0xc001cd8a50, 0x1d1ce80, 0xc0011b1100, 0xc002274a20, 0x0, ...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/plugin.go:80 +0xe3
coredns[23120]: github.com/coredns/coredns/plugin/log.Logger.ServeDNS(0x1cd9c40, 0xc000426f60, 0xc00016e900, 0x1, 0x1, 0x1a76680, 0x2b9c3e0, 0x10, 0x10, 0x1cfd680, ...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/log/log.go:35 +0x384
coredns[23120]: github.com/coredns/coredns/plugin.NextOrFailure(0x19ce11b, 0xa, 0x1cd9d00, 0xc000102ff0, 0x1cfd680, 0xc001cd8a50, 0x1d1ce80, 0xc0011b10c0, 0xc002274a20, 0x0, ...)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/plugin.go:80 +0xe3
coredns[23120]: github.com/coredns/coredns/plugin/metrics.(*Metrics).ServeDNS(0xc000480120, 0x1cfd680, 0xc001cd8a50, 0x1d1cfa0, 0xc001d60a80, 0xc002274a20, 0x40c368, 0x30, 0x18d2200)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/plugin/metrics/handler.go:27 +0x273
coredns[23120]: github.com/coredns/coredns/core/dnsserver.(*Server).ServeDNS(0xc0005da8a0, 0x1cfd680, 0xc001cd8a50, 0x1d1cfa0, 0xc001d60a80, 0xc002274a20)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/core/dnsserver/server.go:242 +0x6d4
coredns[23120]: github.com/coredns/coredns/core/dnsserver.(*Server).ServePacket.func1(0x1d1ed40, 0xc000c54cc0, 0xc002274a20)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/src/github.com/coredns/coredns/core/dnsserver/server.go:115 +0xb0
coredns[23120]: github.com/miekg/dns.HandlerFunc.ServeDNS(0xc00095e220, 0x1d1ed40, 0xc000c54cc0, 0xc002274a20)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/pkg/mod/github.com/miekg/dns@v1.1.14/server.go:37 +0x44
coredns[23120]: github.com/miekg/dns.(*Server).serveDNS(0xc000a4ad00, 0xc0010a5000, 0x27, 0x200, 0xc000c54cc0)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/pkg/mod/github.com/miekg/dns@v1.1.14/server.go:603 +0x2bc
coredns[23120]: github.com/miekg/dns.(*Server).serveUDPPacket(0xc000a4ad00, 0xc000b623d0, 0xc0010a5000, 0x27, 0x200, 0xc000bea100, 0xc0009c18e0)
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/pkg/mod/github.com/miekg/dns@v1.1.14/server.go:549 +0xb2
coredns[23120]: created by github.com/miekg/dns.(*Server).serveUDP
coredns[23120]: /tmp/tmp.nhEpihdcHr/g/pkg/mod/github.com/miekg/dns@v1.1.14/server.go:479 +0x27a

```
the completed dumped file can see the attachment:
[dump-34.txt](https://github.com/coredns/coredns/files/3358210/dump-34.txt)

### PR DESCRIPTION
**Pull request by miekg**
Simplify the locking, remove the reloadMu and just piggyback on the other
lock for accessing content, which assumes things can be move underneath.

Fixes: #2975

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

**Comment by miekg** [comment]
sitll racing:
~~~
WARNING: DATA RACE
Write at 0x00c0002618f0 by goroutine 28:
  github.com/coredns/coredns/plugin/file.(*Zone).Reload.func1()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/reload.go:37 +0x5be
Previous read at 0x00c0002618f0 by goroutine 93:
  github.com/coredns/coredns/plugin/file.(*Zone).Lookup()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/lookup.go:181 +0x31a3
  github.com/coredns/coredns/plugin/file.File.ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/file.go:82 +0xce8
~~~

**Comment by miekg** [comment]
also don't fully trust the call to `additionalProcessing`. In fact I'm pondering the wisdom of having these apex records separately in the first place.

**Comment by miekg** [comment]
this requires some more invasive changes in naming of certain functions (additionalProcessing has two variants) and more helpers can be pushed down to tree or a rrutil helper dir might be required to aid in readability.

File: plugin/file/lookup.go [hunk 1]
```diff
@@ -270,54 +263,50 @@ Out:
 	return nil, ret, nil, rcode
 }
 
-// Return type tp from e and add signatures (if they exists) and do is true.
-func (z *Zone) typeFromElem(elem *tree.Elem, tp uint16, do bool) []dns.RR {
+// typeFromElem returns the type tp from e and adds signatures (if they exists) and do is true.
```
Comment (stp-ip) [inline plugin/file/lookup.go]:
```suggestion
// typeFromElem returns the type tp from e and adds signatures (if they exist) and do is true.
```

**Review by stp-ip — APPROVED** [review]
/lgtm

nit: previous typo

**Pull request merged**